File: src/main/java/net/usbwire/usbplus/util/MultiVersion.java
package net.usbwire.usbplus.util;

import net.minecraft.entity.Entity;
import net.minecraft.item.Item;
import net.minecraft.registry.Registries;
import java.awt.Color;

import net.usbwire.usbplus.interfaces.EntityMixinInterface;
import net.usbwire.usbplus.USBPlus;

/**
 * For when there are differences between versions
 * Inspired by JsMacros
 */
public final class MultiVersion {
	private MultiVersion() {}

	public final class ItemHelper {

		private ItemHelper() {}

		/**
		 * returns item identifier (minecraft:item)
		 */
		public static String getItemId(Item item) {
			return Registries.ITEM.getId(item).toString();
		}
	}

	public final class EntityHelper {

		private EntityHelper() {}

		public static int getGlowingColor(Entity entity) {
			return entity.getTeamColorValue();
		}

		public static void setGlowingColor(Entity entity, Color color) {
			EntityMixinInterface special = (EntityMixinInterface) entity;
			int intColor = color.getRGB();
			if (entity.getTeamColorValue() == intColor)
				return;
			special.usbplus_setGlowingColor(intColor);
		}

		public static void resetGlowingColor(Entity entity) {
			EntityMixinInterface special = (EntityMixinInterface) entity;
			special.usbplus_resetColor();
		}

		public static void setGlowing(Entity entity, boolean value) {
			EntityMixinInterface special = (EntityMixinInterface) entity;
			special.usbplus_setForceGlowing(value ? 2 : 0);
		}

		public static boolean isGlowingForced(Entity entity) {
			EntityMixinInterface special = (EntityMixinInterface) entity;
			return special.usbplus_getForceGlowing() != 1;
		}

		public static void resetGlowing(Entity entity) {
			EntityMixinInterface special = (EntityMixinInterface) entity;
			special.usbplus_setForceGlowing(1);
		}
	}
}

File: src/main/kotlin/net/usbwire/usbplus/util/MultiVersion.kt

import net.minecraft.entity.Entity
import net.minecraft.item.Item
import net.minecraft.registry.Registries
import java.awt.Color

import net.usbwire.usbplus.interfaces.EntityMixinInterface
import net.usbwire.usbplus.USBPlus

/**
 * For when there are differences between versions
 * Inspired by JsMacros
 */
object ItemHelper {

	/**
	 * returns item identifier (minecraft:item)
	 */
	fun getItemId(item: Item?): String {
		return Registries.ITEM.getId(item).toString()
	}
}

object EntityHelper {
	fun getGlowingColor(entity: Entity): Int {
		return entity.teamColorValue
	}

	fun setGlowingColor(entity: Entity, color: Color) {
		val special = entity as EntityMixinInterface
		val intColor = color.rgb
		if (entity.teamColorValue == intColor) return;
		special.usbplus_setGlowingColor(intColor)
	}

	fun resetGlowingColor(entity: Entity) {
		val special = entity as EntityMixinInterface
		special.usbplus_resetColor()
	}

	fun setGlowing(entity: Entity, value: Boolean) {
		val special = entity as EntityMixinInterface
		special.usbplus_setForceGlowing(if (value) 2 else 0)
	}

	fun isGlowingForced(entity: Entity): Boolean {
		val special = entity as EntityMixinInterface
		if (special.usbplus_getForceGlowing() == 1) return false
		return true
	}

	fun resetGlowing(entity: Entity) {
		val special = entity as EntityMixinInterface
		special.usbplus_setForceGlowing(1)
	}
}